### PR TITLE
interpreter,compiler(amd64): complete SIMD instructions

### LINKED ISSUE #210: Document all wazeroir instructions with ref to the spec
docs in instruction.go is half baked currently..

### PR DESCRIPTION
**Pull request by mathetake**
This completes the implementation of SIMD proposal for both 
the interpreter and compiler(amd64).
This also fixes #210 by adding the complete documentation 
over all the wazeroir operations.

arm64 compiler will follow in subsequent PRs.

part of #484

### CODE REVIEW

**Comment by mathetake** [comment]
ok passed `simd_**_arith.wast`, `simd_**_arith2.wast`, `simd_f32.wast`, `simd_f64.wast` spectests with both engines... will implement more!

**Comment by mathetake** [comment]
ok `simd_i16x8_q15mulr_sat_s`, `simd_int_to_int_extend`, `simd_i64x2_extmul_i32x4`, `simd_i32x4_extmul_i16x8` and `simd_i16x8_extmul_i8x16` passed...

**Comment by mathetake** [comment]
ok passed all spec test with the interpreter, meaning that it is fully compatible with WebAssembly 2.0 draft!

**Comment by mathetake** [comment]
ok passed all tests for both engines... backfilling comments and unit tests...

File: internal/engine/compiler/compiler.go [hunk 1]
```diff
@@ -16,474 +16,286 @@ type compiler interface {
...
-	// If "index" >= len(defaults), then branch into the L_DEFAULT label.
-	// Otherwise, we enter label of targets[index].
-	// See wasm.OpcodeBrTable
+	// compileBrTable adds instructions to perform wazeroir.OperationBrTable.
 	compileBrTable(o *wazeroir.OperationBrTable) error
-	// compileCall adds instructions to call into a function of the given index.
-	// See wasm.OpcodeCall
+	// compileCall adds instructions to perform wazeroir.OperationCall.
 	compileCall(o *wazeroir.OperationCall) error
-	// compileCallIndirect adds instructions to perform call_indirect operation.
```
Comment (mathetake) [inline internal/engine/compiler/compiler.go:61]: These docs are migrated to wazeroir/operations.go

**Comment by mathetake** [comment]
ok finished the tests for spectest.go... marked this for ready!

File: internal/asm/amd64/consts.go [hunk 2]
```diff
@@ -323,16 +323,16 @@ const (
 	PADDB
 	// PADDW is the PADDW instruction. https://www.felixcloutier.com/x86/paddb:paddw:paddd:paddq
 	PADDW
-	// PADDL is the PADDD instruction. https://www.felixcloutier.com/x86/paddb:paddw:paddd:paddq
-	PADDL
+	// PADDD is the PADDD instruction. https://www.felixcloutier.com/x86/paddb:paddw:paddd:paddq
+	PADDD
```
Comment (mathetake) [inline internal/asm/amd64/consts.go:327]: fixed the names to match with felix, not with Go asm.

File: internal/asm/amd64/consts.go [hunk 3]
```diff
@@ -323,16 +323,16 @@ const (
...
 	// PADDQ is the PADDQ instruction. https://www.felixcloutier.com/x86/paddb:paddw:paddd:paddq
 	PADDQ
 	// PSUBB is the PSUBB instruction. https://www.felixcloutier.com/x86/psubb:psubw:psubd
 	PSUBB
 	// PSUBW is the PSUBW instruction. https://www.felixcloutier.com/x86/psubb:psubw:psubd
 	PSUBW
-	// PSUBL is the PSUBD instruction. https://www.felixcloutier.com/x86/psubb:psubw:psubd
-	PSUBL
+	// PSUBD is the PSUBD instruction. https://www.felixcloutier.com/x86/psubb:psubw:psubd
+	PSUBD
```
Comment (mathetake) [inline internal/asm/amd64/consts.go:335]: ditto

File: internal/integration_test/spectest/spectest_test.go [hunk 4]
```diff
@@ -0,0 +1,581 @@
+package spectest
```
Comment (mathetake) [inline internal/integration_test/spectest/spectest_test.go:1]: ps I added unit tests of specctest.go! test of test!

File: Makefile [hunk 5]
```diff
@@ -102,7 +102,7 @@ build.spectest.v2: # Note: SIMD cases are placed in the "simd" subdirectory.
 
 .PHONY: test
 test:
-	@go test $$(go list ./... | grep -v spectest) -timeout 120s
+	@go test $$(go list ./... | grep -vE '$(spectest_v1_dir)|$(spectest_v2_dir)') -timeout 120s
```
Comment (mathetake) [inline Makefile:105]: needed to run spectest/spectest_test.go

File: internal/wazeroir/operations.go [hunk 6]
```diff
@@ -2,13 +2,15 @@ package wazeroir
```
Comment (mathetake) [inline internal/wazeroir/operations.go:2]: migrated the docs on engine/compiler/compiler.go here

File: Makefile (hunk 5 above)
Comment (codefromthecrypt) [reply Makefile:105]: cool we can tidy this up in make later if needed.

File: internal/asm/amd64/impl.go [hunk 7]
```diff
@@ -1330,6 +1330,122 @@ var registerToRegisterOpcode = map[asm.Instruction]struct {
...
+	MULPD: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x59}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/divps
+	DIVPS: {opcode: []byte{0x0f, 0x5e}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/divpd
+	DIVPD: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x5e}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/sqrtps
+	SQRTPS: {opcode: []byte{0x0f, 0x51}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/sqrtpd
+	SQRTPD:  {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x51}, requireSrcFloat: true, requireDstFloat: true},
+	ROUNDPS: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x3a, 0x08}, requireSrcFloat: true, requireDstFloat: true, needArg: true},
```
Comment (codefromthecrypt) [inline internal/asm/amd64/impl.go:1389]:
```suggestion
	// https://www.felixcloutier.com/x86/roundps
	ROUNDPS: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x3a, 0x08}, requireSrcFloat: true, requireDstFloat: true, needArg: true},
```

File: internal/asm/amd64/impl.go [hunk 8]
```diff
@@ -1330,6 +1330,122 @@ var registerToRegisterOpcode = map[asm.Instruction]struct {
...
+	ROUNDPD: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x3a, 0x09}, requireSrcFloat: true, requireDstFloat: true, needArg: true},
```
Comment (codefromthecrypt) [inline internal/asm/amd64/impl.go:1390]:
```suggestion
	// https://www.felixcloutier.com/x86/roundpd
	ROUNDPD: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x3a, 0x09}, requireSrcFloat: true, requireDstFloat: true, needArg: true},
```

File: internal/asm/amd64/impl.go [hunk 9]
```diff
@@ -1330,6 +1330,122 @@ var registerToRegisterOpcode = map[asm.Instruction]struct {
...
+	// https://www.felixcloutier.com/x86/punpckhbw:punpckhwd:punpckhdq:punpckhqdq
+	PUNPCKHWD: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x69}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/pmulhuw
+	PMULHUW: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0xe4}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/pmuldq
+	PMULDQ: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x38, 0x28}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/pmulhrsw
+	PMULHRSW: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x38, 0x0b}, requireSrcFloat: true, requireDstFloat: true},
+	// https://www.felixcloutier.com/x86/pmovsx
+	PMOVSXBW: {mandatoryPrefix: 0x66, opcode: []byte{0x0f, 0x38, 0x20}, requireSrcFloat: true, requireDstFloat: true},
```
Comment (codefromthecrypt) [inline internal/asm/amd64/impl.go:1404]: Can you add something like this to `registerToRegisterOpcode` to help next person figure out when an instruction is missing ;) I don't know the correct words, so gave it a shot.

ex.
```go
// The index of felixcloutier.com does not include all instruction variations.
// For example, removing the suffix `BW` from `PMOVSXBW` allows you to find it in https://www.felixcloutier.com/x86/index.html.
//
// Here are the known instruction conventions used in felixcloutier.com
//	* BW - 16-bit integer
//	* BD - 32-bit integer
//	* BQ - 64-bit integer
```

**Review by codefromthecrypt — COMMENTED** [review]
unleashing a few

**Review by codefromthecrypt — COMMENTED** [review]
I notice we are getting to tens of thousands of tests due to matrix `t.Run`

While this is a good default, I think that in the case of instruction boundary testing, especially later with fuzzing, we can do it differently to keep things like IDEs from crashing, and also keep the runtime lower.

Basically substitute dimensional explosions with a format string on input instead of `t.Run`
ex. `require.False(t, x, "input %d", y)`

Then in RATIONALE.md, say...


### Test matrices

Typically, we advice boundary tests to use go test tables, where `t.Run` is run repeatedly with different inputs, using pinned variables. An exception to this are compiler instruction tests, where there are too many inputs, resulting in tens of thousands of tests.

Instead, we run a normal loop substituting dimensional explosions with a format string on input instead of `t.Run`
ex. `require.False(t, x, "input %d", y)`

This keeps the test cardinality low which reduces runtime. On the rare failure, the input is visible in the failure output. While re-running this input is less ideal than a separate test, it is possible manually by temporarily removing the loop.

File: internal/engine/interpreter/interpreter.go [hunk 10]
```diff
@@ -2924,9 +3011,1290 @@ func (ce *callEngine) callNativeFunc(ctx context.Context, callCtx *wasm.CallCont
...
+						if subbed := int64(v) + int64(w); subbed < 0 {
+							uv = uint64(byte(0))
+						} else if subbed > math.MaxUint8 {
+							uv = uint64(byte(0xff))
+						} else {
+							uv = uint64(byte(subbed))
+						}
+					}
+
+					if i < 8 {
```
Comment (codefromthecrypt) [inline internal/engine/interpreter/interpreter.go:3048]: trailing comment (same feedback for below case on i <4)

File: internal/integration_test/spectest/spectest.go [hunk 11]
```diff
@@ -597,52 +644,157 @@ func testdataPath(filename string) string {
...
-			expectedTypesVectorFlattend = append(expectedTypesVectorFlattend, wasm.ValueTypeI64)
-			expectedTypesVectorFlattend = append(expectedTypesVectorFlattend, wasm.ValueTypeI64)
+// valuesEq returns true if all the actual result matches exps which are all expressed as uint64.
+// 	* actual,exps: comparison target values which are all represented as uint64, meaning that if valTypes = [V128,I32], then
+//		we have actual/exp = [(lower-64bit of the first V128), (higher-64bit of the first V128), I32].
+// 	* valTypes holds the wasm.ValueType(s) of the original values in Wasm.
+// 	* laneTypes maps the index of valueTypes to laneType if valueTypes[i] == wasm.ValueTypeV128.
+//
+// Also, if matched == false this returns non-empty valuesMsg which can be used to augment the test failure message.
+func valuesEq(actual, exps []uint64, valTypes []wasm.ValueType, laneTypes map[int]laneType) (matched bool, valuesMsg string) {
```
Comment (codefromthecrypt) [inline internal/integration_test/spectest/spectest.go:654]: wow good improvement

**Review by codefromthecrypt — APPROVED** [review]
this is great, especially as I'm sure arm64 is gonna be easier!

main thing to do now or later is pay down some techdebt accruing in interpreter.go. I feel this part of the codebase may see some contributions by external folks, so it could be a bit more inviting.

**Pull request merged**